### PR TITLE
Improve byte vector to string conversion

### PR DESCRIPTION
**Pull request by powerboat9**

### CODE REVIEW

File: gcc/rust/expand/rust-macro-builtins.cc [hunk 1]
```diff
@@ -576,10 +576,8 @@ MacroBuiltin::include_str_handler (location_t invoc_locus,
   if (expect_single)
     rust_error_at (invoc_locus, "%s was not a valid utf-8 file",
 		   target_filename.c_str ());
-  else if (!bytes.empty ())
-    str = std::string ((const char *) &bytes[0], bytes.size ());
   else
-    return tl::nullopt;
```
Comment (dkm) [inline gcc/rust/expand/rust-macro-builtins.cc:582]: It's not strictly equivalent, as `nullopt` can be used to signal that the transcriber didn't produce anything. Maybe that doesn't really matter, or it's better to have an empty fragment created here with sloc info. @CohenArthur ?

**Review by philberty — COMMENTED** [review]
Do we not need to keep a check for bytes.empty() ?

Maybe if we get bytes.empty we can make str = empty string? and return to::nullopt?

**Comment by dkm** [comment]
> Do we not need to keep a check for bytes.empty() ?
> 
> Maybe if we get bytes.empty we can make str = empty string? and return to::nullopt?

You mean replacing the `if (bytes.empty()) return nullopt;` by a later `if str.empty() return nullopt`?

**Comment by powerboat9** [comment]
Can you not ```include_str!``` an empty file?

**Comment by philberty** [comment]
Good point @powerboat9 I am just thinking its probably not safe to do .data on the bytes vector if its empty. How about we do:

```
if (!bytes.empty ())
    str = std::string ((const char *) &bytes[0], bytes.size ());
else
  str = std::string ();
```

With no null so like you say we handle empty includes

**Comment by powerboat9** [comment]
https://en.cppreference.com/w/cpp/container/vector/data seems to suggest that calling ```data``` on a vector should produce a valid result even if the vector is empty.

**Comment by philberty** [comment]
Good find @powerboat9 lets go for it :D

**Review by philberty — APPROVED** [review]
LGTM

**Comment by philberty** [comment]
lets try this again

**Pull request merged**